src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs

```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                var gcSuppressFinalizeMethodSymbol = compilationContext.Compilation
+                                                        .GetTypeByMetadataName("System.GC")
+                                                        .GetMembers("SuppressFinalize")
+                                                        .OfType<IMethodSymbol>()
+                                                        .SingleOrDefault();
+
+                if (gcSuppressFinalizeMethodSymbol == null)
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationBlockStartAction(operationBlockContext =>
+                {
+                    if (operationBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = (IMethodSymbol)operationBlockContext.OwningSymbol;
+                    if (methodSymbol.IsExtern || methodSymbol.IsAbstract)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new SuppressFinalizeAnalyzer(methodSymbol, gcSuppressFinalizeMethodSymbol, compilationContext.Compilation);
+
+                    operationBlockContext.RegisterOperationAction(analyzer.Analyze, OperationKind.InvocationExpression);
+                    operationBlockContext.RegisterOperationBlockEndAction(analyzer.OperationBlockEndAction);
+                });
+            });
+
+        }
+
+        private class SuppressFinalizeAnalyzer
+        {
+            private enum SuppressFinalizeUsage
+            {
+                CanCall,
+                MustCall,
+                MustNotCall
+            }
+
+            private bool _suppressFinalizeCalled = false;
+
+            private readonly Compilation _compilation;
+            private readonly IMethodSymbol _containingMethodSymbol;
+            private readonly IMethodSymbol _gcSuppressFinalizeMethodSymbol;
+
+            private SemanticModel _semanticModel;
+            private IInvocationExpression _invocationExpression;
+
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, IMethodSymbol gcSuppressFinalizeMethodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+                this._gcSuppressFinalizeMethodSymbol = gcSuppressFinalizeMethodSymbol;
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                _invocationExpression = (IInvocationExpression)analysisContext.Operation;
```

Comment:
You don't need to store the invocationExpression onto a field. I would recommend following change:

1. Make invocationExpression a local.
2. Move the entire block from line 156-181 inside the if block at the end of this method, i.e. line 134 - they should be diagnostics reported on the invocation Operation, not in OperationBlockEndAction.
3. Can you compute and store the semantic model only for the first time, i.e. if it is null?